### PR TITLE
Tokenizing input final

### PR DESCRIPTION
**Pull request by juliachin123**
Everything should be working properly now.

### CODE REVIEW

File: src/codeu/chat/client/commandline/Chat.java [hunk 1]
```diff
@@ -47,9 +47,13 @@ public Chat(Context context) {
   //
   public boolean handleCommand(String line) {
 
-    final Scanner tokens = new Scanner(line.trim());
-
-    final String command = tokens.hasNext() ? tokens.next() : "";
+    final List<String> args = new ArrayList<>();
```
Comment (evelez7) [inline src/codeu/chat/client/commandline/Chat.java]: You deleted the tokens object but tokens is still being used below (on line 73). There will be a complaint at compilation stating that it cannot find symbol `tokens`. 

Also, List has not been imported to this class so there will be a "cannot find symbol error" for List.

File: src/codeu/chat/client/commandline/Chat.java [hunk 2]
```diff
@@ -47,9 +47,13 @@ public Chat(Context context) {
...
+    final Tokenizer tokenizer = new Tokenizer(line);
```
Comment (evelez7) [inline src/codeu/chat/client/commandline/Chat.java]: The Tokenizer class has not been imported so there will be another "cannot find symbol" error.

File: test/codeu/chat/TestRunner.java [hunk 3]
```diff
@@ -28,11 +28,21 @@ public static void main(String[] args) {
              codeu.chat.server.RawControllerTest.class,
              codeu.chat.util.TimeTest.class,
              codeu.chat.util.UuidTest.class,
-             codeu.chat.util.store.StoreTest.class
+             codeu.chat.util.store.StoreTest.class,
```
Comment (evelez7) [inline test/codeu/chat/TestRunner.java:31]: When I try running TestRunner, it says the class can not be found or loaded. Not exactly sure what's causing this.

File: test/codeu/chat/util/TokenizerTest.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		assertEquals(tokenizer.next(), "\"hello");
+		assertEquals(tokenizer.next(), "world");
+		assertEquals(tokenizer.next(), "how");
+		assertEquals(tokenizer.next(), "are");
+		assertEquals(tokenizer.next(), "you");
+		assertEquals(tokenizer.next(), null);
+	}
+	@Test
+	public void testWithWhiteSpaces() throws IOException {
+		final Tokenizer tokenizer = new Tokenizer("  ");
```
Comment (Tuxdude) [inline test/codeu/chat/util/TokenizerTest.java]: Could you also add few `\n`, `\r`, `\t` characters in the input string since they are also whitespaces.

**Review by evelez7 — COMMENTED** [review]
The build will not compile due to missing imports and symbols.

File: test/codeu/chat/util/TokenizerTest.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+		assertEquals(tokenizer.next(), "  ");
+	}
+	@Test
+	public void testWithNoText() throws IOException {
+		final Tokenizer tokenizer = new Tokenizer(null);
```
Comment (Tuxdude) [inline test/codeu/chat/util/TokenizerTest.java]: The input should be `""` and not `null`.

If you want to test for the null case, please add a separate test if needed. I would not worry about testing null specifically here.

File: test/codeu/chat/util/TokenizerTest.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	}
+	@Test
+	public void testWithEmptyString() throws IOException {
+		final Tokenizer tokenizer = new Tokenizer("");
+		assertEquals(tokenizer.next(), "");
+		assertEquals(tokenizer.next(), null);
+	}
+	@Test
+	public void testWithOneQuotation() throws IOException {
+		final Tokenizer tokenizer = new Tokenizer("\"hello world how are you");
```
Comment (Tuxdude) [inline test/codeu/chat/util/TokenizerTest.java]: Could you double check if this is the expected output when you have only one quotation ?

**Review by Tuxdude — CHANGES_REQUESTED** [review]
I'm sorry but your code still does not compile. Did you try building the code and ensuring the tests pass before sending the pull request? Please please do that, that is the only quick and real way to verify the changes.

Also try running `git status` and it will tell you if you have any local files/changes which have not been committed. This could be another cause why you might see the builds succeeding but others still see a failing build.

```
src/codeu/chat/client/commandline/Panel.java:51: error: cannot find symbol
  public boolean handleCommand(String commandName, List<String> args) {
                                                   ^
  symbol:   class List
  location: class Panel
src/codeu/chat/client/commandline/Chat.java:50: error: cannot find symbol
    final List<String> args = new ArrayList<>();
          ^
  symbol:   class List
  location: class Chat
src/codeu/chat/client/commandline/Chat.java:50: error: cannot find symbol
    final List<String> args = new ArrayList<>();
                                  ^
  symbol:   class ArrayList
  location: class Chat
src/codeu/chat/client/commandline/Chat.java:51: error: cannot find symbol
    final Tokenizer tokenizer = new Tokenizer(line);
          ^
  symbol:   class Tokenizer
  location: class Chat
src/codeu/chat/client/commandline/Chat.java:51: error: cannot find symbol
    final Tokenizer tokenizer = new Tokenizer(line);
                                    ^
  symbol:   class Tokenizer
  location: class Chat
src/codeu/chat/client/commandline/Chat.java:73: error: cannot find symbol
    if (panels.peek().handleCommand(command, tokens)) {
                                             ^
  symbol:   variable tokens
  location: class Chat
6 errors
1 warning
Build FAILED
```

File: test/codeu/chat/TestRunner.java (hunk 3 above)
Comment (Tuxdude) [reply test/codeu/chat/TestRunner.java:31]: I think earlier builds were failing and that is a pre-requisite to be able to run the tests.

With recent changes from Julia, I think it runs the tests.

EDIT: This is a reply to Erick's comment itself, dont know why github shows them isolated :|

File: test/codeu/chat/util/TokenizerTest.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+		assertEquals(tokenizer.next(), "\"hello");
+		assertEquals(tokenizer.next(), "world");
+		assertEquals(tokenizer.next(), "how");
+		assertEquals(tokenizer.next(), "are");
+		assertEquals(tokenizer.next(), "you");
+		assertEquals(tokenizer.next(), null);
+	}
+	@Test
+	public void testWithWhiteSpaces() throws IOException {
+		final Tokenizer tokenizer = new Tokenizer("\n \r \t");
```
Comment (Tuxdude) [inline test/codeu/chat/util/TokenizerTest.java:42]: This matches your implementation, but do you really want the whitespace characters as tokens?

If it were me, I would modify Tokenizer to automatically ignore the whitespaces (use `String.trim()` method I suggested last time). It should be fine to do this in a separate pull request as this part is not covered directly by your mini feature.

Also this test is failing, you will still have to fix this.

File: test/codeu/chat/util/TokenizerTest.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+	}
+	@Test
+	public void testWithQuotes() throws IOException {
+		final Tokenizer tokenizer = new Tokenizer("\"hello world\" \"how are you\"");
+		assertEquals(tokenizer.next(), "hello world");
+		assertEquals(tokenizer.next(), "how are you");
+		assertEquals(tokenizer.next(), null);
+	}
+	@Test
+	public void testWithEmptyString() throws IOException {
```
Comment (Tuxdude) [inline test/codeu/chat/util/TokenizerTest.java:26]: This test is also failing.

File: src/codeu/chat/client/commandline/Chat.java [hunk 9]
```diff
@@ -18,6 +18,10 @@
 import java.util.Map;
 import java.util.Scanner;
```
Comment (Tuxdude) [inline src/codeu/chat/client/commandline/Chat.java:19]: It should be safe to get rid of this import.

File: src/codeu/chat/client/commandline/Panel.java [hunk 10]
```diff
@@ -17,6 +17,7 @@
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Scanner;
```
Comment (Tuxdude) [inline src/codeu/chat/client/commandline/Panel.java:19]: It should be safe to get rid of this import as well.

File: test/codeu/chat/TestRunner.java [hunk 11]
```diff
@@ -28,11 +28,21 @@ public static void main(String[] args) {
...
+             codeu.chat.util.TokenizerTest.class
          );
       for (final Failure failure : result.getFailures()) {
          System.out.println(failure.toString());
       }
       System.out.println(result.wasSuccessful());
    }
 }
+
```
Comment (Tuxdude) [inline test/codeu/chat/TestRunner.java:40]: nit: You can get rid of these extra empty lines.

**Review by Tuxdude — CHANGES_REQUESTED** [review]
For the most part changes look good, except for the 2 failing test cases. Please fix them.

https://travis-ci.org/EVelez79/CodeU-ProjectGroup6/jobs/241463325

```
$ python build.py run codeu.chat.TestRunner
Running: [ java -cp bin:third_party/junit4-4.11.jar:third_party/hamcrest-core-1.3.jar codeu.chat.TestRunner ]
testWithWhiteSpaces(codeu.chat.util.TokenizerTest): expected:<null> but was:<
>
testWithEmptyString(codeu.chat.util.TokenizerTest): expected:<null> but was:<>
false
```

**Review by netcables — APPROVED** [review]
LGTM. Tests all passed and all of the commands worked properly when I tried to run ClientMain.

**Comment by Tuxdude** [comment]
LGTM. Please remove the unused Scanner imports in `Chat.java` and `Panel.java` like I suggested earlier.

**Review by Tuxdude — APPROVED** [review]

**Pull request merged**